crates/iota-indexer/src/apis/governance_api.rs

```diff
@@ -375,68 +382,72 @@ impl<T: R2D2Connection + 'static> GovernanceReadApi<T> {
         &self,
         system_state_summary: &IotaSystemStateSummary,
     ) -> Result<Vec<ValidatorExchangeRates>, IndexerError> {
-        // Get validator rate tables
-        let mut tables = vec![];
+        let (active_rates, inactive_rates) = tokio::try_join!(
+            self.active_validators_exchange_rate(system_state_summary),
+            self.inactive_validators_exchange_rate(system_state_summary)
+        )?;
 
-        for validator in &system_state_summary.active_validators {
-            tables.push((
-                validator.iota_address,
-                validator.staking_pool_id,
-                validator.exchange_rates_id,
-                validator.exchange_rates_size,
-                true,
-            ));
-        }
+        Ok(active_rates
+            .into_iter()
+            .chain(inactive_rates.into_iter())
+            .collect())
+    }
 
-        // Get inactive validator rate tables
-        for df in self
-            .inner
-            .get_dynamic_fields_in_blocking_task(
+    /// Check for validators in the `Active` state and get its exchange rate
+    async fn active_validators_exchange_rate(
+        &self,
+        system_state_summary: &IotaSystemStateSummary,
+    ) -> Result<Vec<ValidatorExchangeRates>, IndexerError> {
+        let tables = system_state_summary
+            .active_validators
+            .iter()
+            .map(|validator| {
+                (
+                    validator.iota_address,
+                    validator.staking_pool_id,
+                    validator.exchange_rates_id,
+                    validator.exchange_rates_size,
+                    true,
+                )
+            })
+            .collect();
+
+        self.validator_exchange_rates(tables).await
+    }
+
+    /// Check for validators in the `Inactive` state and get its exchange rate
+    async fn inactive_validators_exchange_rate(
+        &self,
+        system_state_summary: &IotaSystemStateSummary,
+    ) -> Result<Vec<ValidatorExchangeRates>, IndexerError> {
+        let tables = self
+            .validator_summary_from_system_state(
                 system_state_summary.inactive_pools_id,
-                None,
-                system_state_summary.inactive_pools_size as usize,
+                system_state_summary.inactive_pools_size,
+                |df| {
+                    bcs::from_bytes::<ID>(&df.bcs_name).map_err(|e| {
+                        IotaError::ObjectDeserialization {
+                            error: e.to_string(),
+                        }
+                        .into()
+                    })
```

Comment:
Can be simplified through `map_err(IndexerError::from)` (there is a `From<bcs::Error>` implementation).

Reply:
[Fixed](https://github.com/iotaledger/iota/pull/4313/commits/05a02543dfc4b99f24c0ce410aaffc9139e5bf30#diff-324f865071b593cb4ed59dd865b9600fae68065a818e622b14f4f5554812fd98R426)